`as` cast improvements

WaffleLapkin opened the pull request:
Improve the description of `as` casts, see individual commits for more info.

WaffleLapkin committed:
fixup list of valid `as` casts

`&m1 T -> *m2 T` is not specific to `as` casts (it's a coercion!), but
`*[T; N]` -> `*T` is! (previously only the reference source version was
documented)

WaffleLapkin committed:
use footnotes instead of `\*` in `as` description

WaffleLapkin committed:
link from `as` cast table to semantic chapters

WaffleLapkin commented:
Am I reading the CI failure correctly? It seems to imply that footnotes are global and you can have `[^1]` in two different files. Why so?...

ehuss commented:
Because all pages are concatenated together to generate the print page. mdbook currently doesn't have the ability to rewrite the markdown to make the footnotes unique.

ehuss reviewed (approved):
Looks good. Will merge with the footnote labels fixed (probably can pick some short words for those?).

I suppose eventually we should be a little clearer about the normative nature of whether or not a footnote counts.

WaffleLapkin committed:
fixup footnotes

apparently mdbook requires them to have globally unique identifiers
(but they are still rendered as ^1, ^2, ...)

WaffleLapkin commented:
@ehuss fixed 👍🏻

BoxyUwU commented on src/expressions/operator-expr.md:
```diff
@@ -454,35 +454,35 @@ Any cast that does not fit either a coercion rule or an entry in the table is a
+| [Function pointer]    | Integer               | Function pointer to address cast                      |
+| Closure [^no-capture] | Function pointer      | Closure to function pointer cast                      |
+
+[^meta-compat]: where `T` and `V` have compatible metadata:
```
The little footnote indicator is also just really easy to miss as part of a big table of information. (But even if I had noticed it I would not have clicked on it to try and find this information)

BoxyUwU replied:
Honestly I could still imagine missing it, when I see that table my first instinct is to click on the "Pointer to pointer casts" header to see *all* information about pointer to pointer casts. Having the footnotes right below would definitely be an improvement though
